### PR TITLE
feat(source): add source connector

### PR DESCRIPTION
**Pull request by donatorsky**
### Description

Adds blob source handling to Azure Storage connector.

### CODE REVIEW

File: source/source_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		require.NoError(t, src.Ack(ctx, record2.Position))
+
+		// No third record available, backoff
+		// By this time, combined iterator should switch from snapshot to CDC iterator
+		record3, err := src.Read(ctx)
+		require.Equal(t, sdk.Record{}, record3)
+		require.ErrorIs(t, err, sdk.ErrBackoffRetry)
+	})
+
+	t.Run("Secondly, CDC iterator runs and reads all blobs created meanwhile", func(t *testing.T) {
```
Comment (maha-hajja) [inline source/source_integration_test.go:95]: add more integration tests for edge case scenarios
ex: like invalid config, empty storage, stop the connector after snapshot and start it again ..

File: source/iterator/snapshot.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		select {
+		case <-w.tomb.Dying():
+			return w.tomb.Err()
+
+		default:
+			if w.paginator.NextPage(w.tomb.Context(context.TODO())) {
+				resp := w.paginator.PageResponse()
+
+				for _, item := range resp.Segment.BlobItems {
+					// Reject item when it wasn't modified since the last iteration
```
Comment (maha-hajja) [inline source/iterator/snapshot.go:99]: why is this in the snapshot iterator? isn't it supposed to be for CDC?

File: README.md [hunk 3]
```diff
@@ -1 +1,50 @@
...
+### Configuration Options
+
+| name               | description                                                                                                                                  | required | default |
+|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------|----------|---------|
+| `connectionString` | Azure Storage connection string as described here: https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string | `true`   |         |
+| `containerName`    | The name of the container to monitor.                                                                                                        | `true`   |         |
+| `pollingPeriod`    | The polling period for the CDC mode, formatted as a time.Duration string. Must be greater then `0`.                                          | `false`  | `1s`    |
+| `maxResults`       | The maximum number of items, per page, when reading container's items. The minimum value is `1`, maximum value is `5000`.                    | `false`  | `5000`  |
+
+## Destination
```
[the remark this answers is not in the document]
Comment (donatorsky) [reply README.md:37]: Gotcha, will apply this to Salesforce and Elasticsearch connectors as well 🙂

File: README.md [hunk 4]
```diff
@@ -1 +1,50 @@
...
+
+### Configuration Options
+
+| name               | description                                                                                                                                  | required | default |
+|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------|----------|---------|
+| `connectionString` | Azure Storage connection string as described here: https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string | `true`   |         |
+| `containerName`    | The name of the container to monitor.                                                                                                        | `true`   |         |
+| `pollingPeriod`    | The polling period for the CDC mode, formatted as a time.Duration string. Must be greater then `0`.                                          | `false`  | `1s`    |
+| `maxResults`       | The maximum number of items, per page, when reading container's items. The minimum value is `1`, maximum value is `5000`.                    | `false`  | `5000`  |
```
[the remark this answers is not in the document]
Comment (donatorsky) [reply README.md:35]: Also updated for Salesforce and Elasticsearch connectors.

File: source/source.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	// E.g.: goroutine stopped and w.tomb.Err() returned empty record and nil error
+	if reflect.DeepEqual(record, sdk.Record{}) {
+		return sdk.Record{}, sdk.ErrBackoffRetry
+	}
+
+	return record, nil
+}
+
+func (s *Source) Ack(_ context.Context, _ sdk.Position) error {
+	return nil // no ack needed
```
[the remark this answers is not in the document]
Comment (donatorsky) [reply source/source.go:114]: Did the same for Salesforce connector.

File: source/iterator/snapshot.go (hunk 2 above)
Comment (donatorsky) [reply source/iterator/snapshot.go]: Thanks, it was wrong comment here. Updated 🙂

File: source/source_integration_test.go (hunk 1 above)
Comment (donatorsky) [reply source/source_integration_test.go:191]: Updated: added new tests

File: README.md [hunk 6]
```diff
@@ -1 +1,47 @@
...
+## General
+
+The Azure Storage plugin is one of [Conduit](https://github.com/ConduitIO/conduit) plugins.
+It currently provides only source connector.
+
+## How to build it
+
+Run `make`.
+
+## Source
```
Comment (hariso) [inline README.md:12]: It would be good to add more detailed info about how snapshot and CDC work. E.g. what happens when a snapshot is stopped, how does CDC detect changes, does it always pull all objects etc.

File: source/iterator/cdc.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+			// Prepare the storage iterator
+			blobListPager := w.client.ListBlobsFlat(&azblob.ContainerListBlobsFlatOptions{
+				Marker:     w.nextKeyMarker,
+				MaxResults: &w.maxResults,
+				Include: []azblob.ListBlobsIncludeItem{
+					azblob.ListBlobsIncludeItemDeleted,
+				},
+			})
+
+			for blobListPager.NextPage(w.tomb.Context(context.TODO())) {
```
Comment (hariso) [inline source/iterator/cdc.go:117]: Should we replace context.TODO()? Also, does this mean that every pollingPeriod, we pull all object metadata from Azure Storage?

File: README.md [hunk 8]
```diff
@@ -1 +1,47 @@
...
+
+The Source connector monitors given Azure Blob container for file changes and generates appropriate Record when change is detected.
+It supports two reading modes and switches them automatically:
+- Snapshot
+- CDC
+
+In **Snapshot mode**, connector reads the current state of the container, meaning it does not include changes made during this process.
```
[the remark this answers is not in the document]
Comment (donatorsky) [reply README.md:19]: If snapshot iterator gets interrupted, it starts over again. Works the same as S3 snapshot iterator, I believe.

When switched to CDC iterator, it will detect all changes made while snapshot iterator was active.

File: source/position/position.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		Key:       key,
+		Timestamp: timestamp,
+		Type:      TypeCDC,
+	}
+}
+
+// Position represents blob item position metadata.
+type Position struct {
+	Key       string
+	Timestamp time.Time
```
[the remark this answers is not in the document]
Comment (donatorsky) [reply source/position/position.go:78]: Sure, I'm adding that in 😉

File: source/position/position.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+// ToRecordPosition converts Position into sdk.Position.
+func (p Position) ToRecordPosition() (sdk.Position, error) {
+	var buffer bytes.Buffer
+
+	if err := gob.NewEncoder(&buffer).Encode(p); err != nil {
+		return nil, err
+	}
+
+	return buffer.Bytes(), nil
```
[the remark this answers is not in the document]
Comment (donatorsky) [reply source/position/position.go:90]: gob includes the information about types, so I feel it fits better for that. Also, it seems to be slightly faster when benchmarked. But besides that there are no other reasons, can be successfully replaced with JSON.

File: source/iterator/cdc.go (hunk 7 above)
Comment (donatorsky) [reply source/iterator/cdc.go:117]: Changed it to `context.Background()`.

Yes, I'm paginating over container blobs with metadata every pollingPeriod.

File: README.md (hunk 6 above)
Comment (donatorsky) [reply README.md:12]: Updated with additional information: 856eea2

File: source/position/position.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hariso) [reply source/position/position.go:95]: Gotcha, thanks for the info. :+1:

**Review by maha-hajja — APPROVED** [review]
